### DISCUSSION TITLE
想了解下MVVM开发过程中对UserControl的交互，或者说打开一个带参数的窗口应该怎么处理。

### DISCUSSION THREAD

**Discussion by base-linsyd** [BYJRK/DotNet-Discussions#79] [MVVM]
最近在学习wpf相关的内容，
然后我想在主窗口里面给usercontrol赋值，因为数据是动态的点击按钮的时候直接通过behavior来操作usercontrol并且初始化一个usercontrol合适吗，其次就是涉及到这个usercontrol的viewmodel应该怎么初始化呢，这个有点迷糊

**Comment by BYJRK** [comment]
一般有这么几种思路：

1. 用 IoC 注册 ViewModel
2. 自己写一个 ViewModelLocator 之类的 Service
3. 直接把 UserControl 的 VM 写在 MainWindow 的 VM 中然后绑定 DataContext
4. 用静态变量之类的方式存储你要传的值
5. 用 Messenger 将需要传的值传给 UserControl

**Reply by base-linsyd** [reply]
okk，谢谢谢谢谢谢谢谢。
我尝试下，现在自己玩这个功能都能实现，只是在mvvm这个模式下做感觉需要抽象出来好多东西，脑袋转不过来

**Comment by weiyoucheng** [comment]
public partial class MenuEditWindowViewModel : ObservableObject {

        [ObservableProperty]
        public Menu _Menu = null!;

        private void Update() {
            //修改菜单方法
        }
    }

    public partial class MenuEditWindow : Window {
        public MenuEditWindowViewModel ViewModel { get; }

        public MenuEditWindow(MenuEditWindowViewModel viewModel)
        {
            ViewModel = viewModel;
        }
    }

    public partial class MainWinViewModel:ObservableObject {
        public Action<Menu>? EditMenuHandler { get; set; }

        /// <summary>
        /// 绑定选择的菜单
        /// </summary>
        [ObservableProperty]
        public Menu _Menu = null!;

        public MainWinViewModel() {

        }

        [RelayCommand]
        private void EditMenu() {
            EditMenuHandler?.Invoke(Menu);
        }
    }

    public partial class MainWin:Window {
        public MainWinViewModel ViewModel { get; }
        public MainWin(MainWinViewModel viewModel) {
            ViewModel = viewModel;
            ViewModel.EditMenuHandler += UpdateMenu;
        }

        private void UpdateMenu(Menu menu) {
            var viewModel = new MenuEditWindowViewModel() { Menu = menu };
            var window = new MenuEditWindow(viewModel);
            window.Owner = this;
            window.ShowDialog();
        }
    }
